### ISSUE TITLE
[FR] builtin `assert`, `assertOrTrace`, `assertAndContinue` and `assertAndContinueOrTrace`

### ISSUE THREAD

**Issue by michele-nuzzi** [IntersectMBO/plutus#6598]
### Describe the feature you'd like

since plutus V3 scripts are forced to return a `unit` to be considered succesful, and in reality this was already considered standard for plutus V1 and V2, but not enforced.

Despite UPLC being functional, there are 2 side effects that can be triggered: tracing and throwing an error.

on a higher level, most languages will instead allow to write contracts that return booleans, and even in the languages who don't the developer will write the logic using booleans and finally convert to a unit.

today to convert from booleans to unit, a function like the one below is used (`ifThenElse` is a variable, assuming the compiler hoisted the forced builtin)

```
(lam bool_var
  (force 
    [
      [
        [
          ifThenElse 
          bool_var
        ] 
        (delay (con unit ()))
      ] 
      (delay (error))
    ]
  )
)
```

because this is such a common operation a few new builtins could be introduced

```haskell
assert ::  bool -> unit
assertOrTrace :: string -> bool -> unit
assertAndContinue :: bool -> a -> a
assertAndContinueOrTrace :: string -> bool -> a -> a
```
`assert` would act as the uplc function above

`OrTrace` variants would take a string as first parameter, potentially to take advantage of partial function applications, and in case of error it would add the string to the traces of the contract exectuting before failing.

`AndContinue` variants are polymorphic in the return type, so that these can be used to assert properties before execution of other parts of the contract

for example

```
(force
  [
    [
      (force (builtin assertAndContinue))
      someCondition
    ]
    (delay
      continuation_body
    )
  ]
)
```

if the implementation of `AndContinue` variants would make it so that the polymorphic argument is not strictly evaluated by default (given the nature of the builtin) it would make it an even better tool!

for example, the uplc below would behave the same exact way of the uplc above:
```
[
  [
    (force (builtin assertAndContinue))
    someCondition
  ]
  continuation_body
]
```

**Comment by effectfully** [comment]
> `assert :: bool -> unit`

Not sure how I feel about this one. On the one hand, yes, it's a sensible builtin and it would save us some. On the other hand it'd only save a constant amount of cost (CPU/MEM/size) per execution, so it's not very significant. [On the gripping hand](http://www.catb.org/jargon/html/O/on-the-gripping-hand.html), a constant overhead per contract execution is linear overhead on the chain, so maybe it's worth it.

> `assertOrTrace :: string -> bool -> unit`
> `assertAndContinueOrTrace :: string -> bool -> a -> a`

I'm not convinced that it's worth adding these. Maybe we could just make `trace ""` not emit anything so that `trace (ifThenElse b "" msg)` doesn't emit anything  when `b` is `True` (and we can erase it as per normal before submitting to the chain). I'd rather that than have two more builtins and a separate optimization pass replacing `assertOrTrace` with `assert` and `assertAndContinueOrTrace` with `assertAndContinue`.

> `assertAndContinue :: bool -> a -> a`

That may be handy...

I'm working on something that is related to this issue, I'll create a new issue soon and we'll see if it can address your points. Meanwhile, I'm going to claim that this is fairly low priority.

**Comment by michele-nuzzi** [comment]
yes `OrTrace` variants would mainly be used in debug I would assume, so it makes sense to exclude them.

I agree this is low priority
